File: sdks/python/src/verdocs/validators.py
"""Field-input validators and fill checks (js-sdk: Templates/Validators.ts, Envelopes/Fields.ts).

These run entirely client-side and never call the API. The regexes are ported
byte-for-byte from the js-sdk and evaluated with search() to match JS
RegExp.test semantics: patterns without anchors match anywhere in the string.
Validators always check strings, because that is all a user can enter in an
HTML input field, and they never raise: they just return a boolean.
"""

from __future__ import annotations

import re
from typing import TypedDict

from .models import EnvelopeField, Role

_EMAIL_REGEX = re.compile(
    r'^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))'
    r"@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$"
)

# The js-sdk credits https://www.regextester.com/1978 for this pattern. re.ASCII pins \d to [0-9],
# matching the JS \d, so unicode digits do not pass where the js-sdk rejects them.
_PHONE_REGEX = re.compile(
    r"((?:\+|00)[17](?: |\-)?|(?:\+|00)[1-9]\d{0,2}(?: |\-)?|(?:\+|00)1\-\d{3}(?: |\-)?)?"
    r"(0\d|\([0-9]{3}\)|[1-9]{0,3})"
    r"(?:((?: |\-)[0-9]{2}){4}|((?:[0-9]{2}){4})|((?: |\-)[0-9]{3}(?: |\-)[0-9]{4})|([0-9]{7}))",
    re.ASCII,
)

_URL_REGEX = re.compile(
    r"https?://(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)"
)

_POSTAL_CODE_REGEX = re.compile(r"^[A-Za-z0-9-\s]{3,10}$")

_NUMBER_REGEX = re.compile(r"^\d+$", re.ASCII)

# Ported verbatim, including the js-sdk's anchoring quirk: the ^ binds only to the first alternative
# and the $ only to the second, so trailing garbage after a YYYY-MM-DD date (or leading garbage before
# a MM-DD-YYYY one) still matches.
_DATE_REGEX = re.compile(r"^(\d{4}[-/]\d{2}[-/]\d{2})|(\d{2}[-/]\d{2}[-/]\d{4})$", re.ASCII)


class _Validator(TypedDict):
    regex: re.Pattern[str]
    label: str


_VALIDATORS: dict[str, _Validator] = {
    "email": {"regex": _EMAIL_REGEX, "label": "Email Address"},
    "phone": {"regex": _PHONE_REGEX, "label": "Phone Number"},
    "url": {"regex": _URL_REGEX, "label": "URL"},
    "postal_code": {"regex": _POSTAL_CODE_REGEX, "label": "Zip/Postal Code"},
    "number": {"regex": _NUMBER_REGEX, "label": "Number"},
    "date": {"regex": _DATE_REGEX, "label": "Date"},
}

_TAG_REGEX = re.compile(r"^[a-zA-Z0-9-]{0,32}$")


def is_valid_input(value: str, validator: str) -> bool:
    """Check a value against one of the named field validators.

    Ports js-sdk Templates/Validators.ts isValidInput.

    Args:
        value: The value to check.
        validator: The validator name; see get_validators for the options.

    Returns:
        True when the validator exists and the value passes it.
    """
    entry = _VALIDATORS.get(validator)
    return entry is not None and entry["regex"].search(value) is not None


def get_validators() -> list[str]:
    """Get the list of available validators for field inputs.

    Ports js-sdk Templates/Validators.ts getValidators.

    Returns:
        The validator names accepted by is_valid_input.
    """
    return list(_VALIDATORS)


def is_valid_email(email: str | None) -> bool:
    """Check whether a string looks like an email address.

    Ports js-sdk Templates/Validators.ts isValidEmail.

    Args:
        email: The value to check; None and the empty string fail.

    Returns:
        True when the value passes the email pattern.
    """
    return bool(email) and _EMAIL_REGEX.search(email) is not None


def is_valid_phone(phone: str | None) -> bool:
    """Check whether a string looks like a phone number.

    Ports js-sdk Templates/Validators.ts isValidPhone.

    Args:
        phone: The value to check; None and the empty string fail.

    Returns:
        True when the value passes the phone pattern.
    """
    return bool(phone) and _PHONE_REGEX.search(phone) is not None


def is_valid_role_name(value: str, roles: list[Role]) -> bool:
    """Check whether a value names one of the template's roles.

    Ports js-sdk Templates/Validators.ts isValidRoleName.

    Args:
        value: The role name to look for.
        roles: The template's roles.

    Returns:
        True when a role with that exact name exists.
    """
    return any(role.name == value for role in roles)


def is_valid_tag(value: str, tags: list[str]) -> bool:
    """Check whether a value is usable as a template tag.

    Ports js-sdk Templates/Validators.ts isValidTag: the value must be 0-32
    alphanumeric-or-dash characters (yes, the empty string passes; ported
    as-is), or already present in the known tag list.

    Args:
        value: The tag to check.
        tags: Already-known tags, which pass regardless of format.

    Returns:
        True when the tag is well formed or already known.
    """
    return _TAG_REGEX.search(value) is not None or value in tags


# Envelopes/Fields.ts


def is_field_filled(field: EnvelopeField, all_recipient_fields: list[EnvelopeField]) -> bool:
    """Check whether an envelope field has been filled in.

    Ports js-sdk Envelopes/Fields.ts isFieldFilled. Text fields honor their
    email/phone validator; signature-like fields expect their marker values
    (signed, initialed, attached); timestamps count as filled because they are
    stamped automatically at submit; radio fields with a group count as filled
    when any field in the group is selected. One ported quirk: a dropdown with
    a null value counts as filled, because the js-sdk only treats the empty
    string as unfilled there.

    Args:
        field: The field to check.
        all_recipient_fields: Every field belonging to the same recipient,
            used to resolve radio groups.

    Returns:
        True when the field holds a submitted value.
    """
    value = field.value
    match field.type:
        case "textarea" | "textbox":
            match field.validator or "":
                case "email":
                    return bool(value) and is_valid_input(value, "email")
                case "phone":
                    return bool(value) and is_valid_input(value, "phone")
                case _:
                    return (value or "").strip() != ""

        case "signature":
            return value == "signed"

        case "initial":
            return value == "initialed"

        # Timestamp fields get automatically filled when the envelope is submitted.
        case "timestamp":
            return True

        case "date":
            return bool(value)

        case "attachment":
            return value == "attached"

        case "dropdown":
            return value != ""

        case "checkbox":
            return value == "true"

        case "radio":
            if field.group:
                return any(f.value == "true" for f in all_recipient_fields if f.group == field.group)

            return field.value == "true"

        case _:
            return False


def is_field_valid(field: EnvelopeField, all_recipient_fields: list[EnvelopeField]) -> bool:
    """Check whether an envelope field is valid for submission.

    Ports js-sdk Envelopes/Fields.ts isFieldValid: optional fields always
    pass, required fields must be filled. The js-sdk carries an open TODO to
    only let optional fields skip validation when they are empty; we mirror
    the current behavior, so a badly-formatted optional field still passes.

    Args:
        field: The field to check.
        all_recipient_fields: Every field belonging to the same recipient,
            used to resolve radio groups.

    Returns:
        True when the field may be submitted.
    """
    return not field.required or is_field_filled(field, all_recipient_fields)

File: sdks/csharp/src/Verdocs.Sdk/Helpers/Validators.cs
using System.Text.RegularExpressions;
using Verdocs.Models;

namespace Verdocs.Helpers;

/// <summary>
/// Field-input validators and field fill/validity checks, ported from the js-sdk's
/// Templates/Validators.ts and Envelopes/Fields.ts. Validators always check strings, because
/// that is all a user can enter in an input field, and they never throw; they just return
/// whether the value is valid.
/// </summary>
public static class Validators
{
    // The patterns are ported verbatim from the js-sdk. RegexOptions.ECMAScript keeps \d, \w,
    // \s, and \b at their ASCII JavaScript semantics so the SDKs agree on what validates.
    private static readonly Regex EmailRegex = new(
        @"^(([^<>()\[\]\\.,;:\s@""]+(\.[^<>()\[\]\\.,;:\s@""]+)*)|("".+""))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$",
        RegexOptions.ECMAScript);

    // @see https://www.regextester.com/1978. Unanchored like the js-sdk original, so a phone
    // number embedded anywhere in the value matches.
    private static readonly Regex PhoneRegex = new(
        @"((?:\+|00)[17](?: |\-)?|(?:\+|00)[1-9]\d{0,2}(?: |\-)?|(?:\+|00)1\-\d{3}(?: |\-)?)?(0\d|\([0-9]{3}\)|[1-9]{0,3})(?:((?: |\-)[0-9]{2}){4}|((?:[0-9]{2}){4})|((?: |\-)[0-9]{3}(?: |\-)[0-9]{4})|([0-9]{7}))",
        RegexOptions.ECMAScript);

    private static readonly Regex UrlRegex = new(
        @"https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)",
        RegexOptions.ECMAScript);

    private static readonly Regex PostalCodeRegex = new(
        @"^[A-Za-z0-9-\s]{3,10}$",
        RegexOptions.ECMAScript);

    private static readonly Regex NumberRegex = new(
        @"^\d+$",
        RegexOptions.ECMAScript);

    // The alternation binds loosely, like the js-sdk original: the first alternative is only
    // anchored at the start and the second only at the end, so trailing garbage after a
    // yyyy-mm-dd date (or leading garbage before a mm-dd-yyyy one) still matches.
    private static readonly Regex DateRegex = new(
        @"^(\d{4}[-\/]\d{2}[-\/]\d{2})|(\d{2}[-\/]\d{2}[-\/]\d{4})$",
        RegexOptions.ECMAScript);

    // {0,32} means the empty string is a valid tag, like the js-sdk original.
    private static readonly Regex TagRegex = new(
        @"^[a-zA-Z0-9-]{0,32}$",
        RegexOptions.ECMAScript);

    // Insertion-ordered to match the js-sdk's Object.keys(VALIDATORS) enumeration.
    private static readonly string[] ValidatorNames = ["email", "phone", "url", "postal_code", "number", "date"];

    private static readonly Dictionary<string, Regex> ValidatorsByName = new()
    {
        ["email"] = EmailRegex,
        ["phone"] = PhoneRegex,
        ["url"] = UrlRegex,
        ["postal_code"] = PostalCodeRegex,
        ["number"] = NumberRegex,
        ["date"] = DateRegex,
    };

    /// <summary>
    /// True when the value passes the named validator. Ports the js-sdk's isValidInput
    /// (Templates/Validators.ts). Unknown validator names always fail.
    /// </summary>
    /// <param name="value">The value to check.</param>
    /// <param name="validator">The validator name; see <see cref="GetValidators"/> for the known names.</param>
    /// <returns>True when the value is valid for the named validator.</returns>
    /// <sdkOperation>validator.isValidInput</sdkOperation>
    /// <sdkGroup>Validators</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static bool IsValidInput(string value, string validator) =>
        ValidatorsByName.TryGetValue(validator, out var regex) && regex.IsMatch(value);

    /// <summary>
    /// The available validator names for field inputs. Ports the js-sdk's getValidators
    /// (Templates/Validators.ts).
    /// </summary>
    /// <returns>The validator names, in the js-sdk's order.</returns>
    /// <sdkOperation>validator.getValidators</sdkOperation>
    /// <sdkGroup>Validators</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static IReadOnlyList<string> GetValidators() => ValidatorNames;

    /// <summary>
    /// True when the value is a well-formed email address. Ports the js-sdk's isValidEmail
    /// (Templates/Validators.ts). Null and empty values fail.
    /// </summary>
    /// <param name="email">The value to check, or null.</param>
    /// <returns>True when the value is a valid email address.</returns>
    /// <sdkOperation>validator.isValidEmail</sdkOperation>
    /// <sdkGroup>Validators</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static bool IsValidEmail(string? email) =>
        !string.IsNullOrEmpty(email) && EmailRegex.IsMatch(email);

    /// <summary>
    /// True when the value contains a plausible phone number. Ports the js-sdk's isValidPhone
    /// (Templates/Validators.ts). Null and empty values fail.
    /// </summary>
    /// <param name="phone">The value to check, or null.</param>
    /// <returns>True when the value contains a valid phone number.</returns>
    /// <sdkOperation>validator.isValidPhone</sdkOperation>
    /// <sdkGroup>Validators</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static bool IsValidPhone(string? phone) =>
        !string.IsNullOrEmpty(phone) && PhoneRegex.IsMatch(phone);

    /// <summary>
    /// True when the value names one of the given template roles. Ports the js-sdk's
    /// isValidRoleName (Templates/Validators.ts).
    /// </summary>
    /// <param name="value">The role name to look for.</param>
    /// <param name="roles">The template's roles.</param>
    /// <returns>True when a role with that exact name exists.</returns>
    /// <sdkOperation>validator.isValidRoleName</sdkOperation>
    /// <sdkGroup>Validators</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static bool IsValidRoleName(string value, IEnumerable<Role> roles) =>
        roles.Any(role => role.Name == value);

    /// <summary>
    /// True when the value is an acceptable tag: up to 32 letters, digits, or hyphens, or an
    /// exact match for an existing tag. Ports the js-sdk's isValidTag
    /// (Templates/Validators.ts), including its quirk that the empty string passes.
    /// </summary>
    /// <param name="value">The tag to check.</param>
    /// <param name="tags">The existing tags.</param>
    /// <returns>True when the tag is acceptable.</returns>
    /// <sdkOperation>validator.isValidTag</sdkOperation>
    /// <sdkGroup>Validators</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static bool IsValidTag(string value, IEnumerable<string> tags) =>
        TagRegex.IsMatch(value) || tags.Contains(value);

    /// <summary>
    /// True when the field has been filled in, per its type's rules: text fields must have a
    /// non-blank (and, with a validator set, valid) value, signature/initial/attachment
    /// fields must have been acted on, timestamps always count as filled, and grouped radio
    /// buttons count when any field in the group is selected. Ports the js-sdk's
    /// isFieldFilled (Envelopes/Fields.ts). Unknown field types are never filled.
    /// </summary>
    /// <param name="field">The field to check.</param>
    /// <param name="allRecipientFields">Every field assigned to the same recipient, used to resolve grouped fields.</param>
    /// <returns>True when the field counts as filled.</returns>
    /// <sdkOperation>envelope.isFieldFilled</sdkOperation>
    /// <sdkGroup>Envelope</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static bool IsFieldFilled(EnvelopeField field, IReadOnlyList<EnvelopeField> allRecipientFields)
    {
        // The js-sdk's destructuring default only fires for undefined, so a JSON-null value
        // takes a different path there in a couple of branches. C# cannot tell null from
        // missing, so both normalize to empty here.
        var value = field.Value ?? string.Empty;
        switch (field.Type)
        {
            case FieldType.Textarea:
            case FieldType.Textbox:
                return (field.Validator ?? string.Empty) switch
                {
                    "email" => value.Length > 0 && IsValidInput(value, "email"),
                    "phone" => value.Length > 0 && IsValidInput(value, "phone"),
                    _ => value.Trim().Length > 0,
                };

            case FieldType.Signature:
                return value == "signed";

            case FieldType.Initial:
                return value == "initialed";

            // Timestamp fields get automatically filled when the envelope is submitted.
            case FieldType.Timestamp:
                return true;

            case FieldType.Date:
                return value.Length > 0;

            case FieldType.Attachment:
                return value == "attached";

            case FieldType.Dropdown:
                return value.Length > 0;

            case FieldType.Checkbox:
                return value == "true";

            case FieldType.Radio:
                if (!string.IsNullOrEmpty(field.Group))
                {
                    return allRecipientFields
                        .Where(f => f.Group == field.Group)
                        .Any(f => f.Value == "true");
                }

                return field.Value == "true";

            default:
                return false;
        }
    }

    // TODO carried from the js-sdk: only let !required bypass validation when the field is
    // empty; today an optional field with an invalid value still passes.
    /// <summary>
    /// True when the field is valid to submit: optional fields always pass, required fields
    /// must be filled. Ports the js-sdk's isFieldValid (Envelopes/Fields.ts).
    /// </summary>
    /// <param name="field">The field to check.</param>
    /// <param name="allRecipientFields">Every field assigned to the same recipient, used to resolve grouped fields.</param>
    /// <returns>True when the field is valid.</returns>
    /// <sdkOperation>envelope.isFieldValid</sdkOperation>
    /// <sdkGroup>Envelope</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static bool IsFieldValid(EnvelopeField field, IReadOnlyList<EnvelopeField> allRecipientFields) =>
        field.Required != true || IsFieldFilled(field, allRecipientFields);
}
